### PR TITLE
Fix the incremental reflow behavior of text-overflow

### LINKED ISSUE #14952: text-overflow: ellipsis - after the ellipsis character is added, it's never removed
STR:
1. Open this URL in Servo:
```
data:text/html,<div style='overflow:hidden;text-overflow:ellipsis'>helloworldhelloworld</div>
```
2. Resize the window to minimum width, so that the ellipsis kicks in. Several times, as the window is resized.
3. Resize the window back to original width.

Expected: ellipsis character eventually disappears, as there is no overflow any more.
Actual: several ellipsis characters are added and remain visible when the window width increases. The text is irreversibly truncated.

### PR DESCRIPTION
**Pull request by notriddle**
This patch allows Servo to incrementally reflow truncated fragments correctly.

* The untruncated version of a fragment is preserved, and when incrementally reflowing, the untruncated version is what gets reflowed. If it needs truncated, it will get truncated again.
* The ellipsis fragments are skipped when incrementally reflowing a line. If it is still needed, it will be recreated.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14952
- [X] There are tests for these changes OR

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14989)

### CODE REVIEW

File: tests/wpt/mozilla/tests/css/text_overflow_reflow.html [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+<html id=html class="reftest-wait">
+<meta charset="utf-8">
+<title>Text overflow should disappear when the container becomes large enough</title>
+<meta name="description" content="This test is targetted at a bug in Servo's incremental reflow engine.">
+<link rel="match" href="text_overflow_reflow_ref.html">
+<style>html{font-size:20px;font-family:Ahem}</style>
+<div id=goat style="width:100px"><p style="text-overflow:ellipsis;overflow:hidden">XXXXXXXXXX</p></div>
+<script>
+var goat = document.getElementById("goat");
+var html = document.getElementById("html");
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/text_overflow_reflow.html]: nit: Maybe just `document.documentElement`?

File: tests/wpt/mozilla/tests/css/text_overflow_reflow.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<!doctype html>
+<html id=html class="reftest-wait">
+<meta charset="utf-8">
+<title>Text overflow should disappear when the container becomes large enough</title>
+<meta name="description" content="This test is targetted at a bug in Servo's incremental reflow engine.">
+<link rel="match" href="text_overflow_reflow_ref.html">
+<style>html{font-size:20px;font-family:Ahem}</style>
+<div id=goat style="width:100px"><p style="text-overflow:ellipsis;overflow:hidden">XXXXXXXXXX</p></div>
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/text_overflow_reflow.html]: Maybe the purpose of the test is clearer with a `em`-based inline size?

File: tests/wpt/mozilla/tests/css/text_overflow_reflow.html [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<!doctype html>
+<html id=html class="reftest-wait">
+<meta charset="utf-8">
+<title>Text overflow should disappear when the container becomes large enough</title>
+<meta name="description" content="This test is targetted at a bug in Servo's incremental reflow engine.">
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/text_overflow_reflow.html]: Given you're writing a description, probably worth mentioning #14952 :)

File: components/layout/fragment.rs [hunk 4]
```diff
@@ -186,6 +186,8 @@ pub enum SpecificFragmentInfo {
     Multicol,
     MulticolColumn,
     UnscannedText(Box<UnscannedTextFragmentInfo>),
+
```
Comment (emilio) [inline components/layout/fragment.rs]: It could be worth either writing down here how these are used, or maybe renaming it to `TruncatedTextFragment`?

**Review by emilio — COMMENTED** [review]
Looks good to me, mostly minor comments :)

File: components/layout/fragment.rs (hunk 4 above)
Comment (notriddle) [inline components/layout/fragment.rs]: It's not necessarily text, though. Non-text fragments get truncated, but in an all-or-nothing fashion.

**Comment by notriddle** [comment]
All the nits should be resolved, now.

@bors-servo try

**Comment by notriddle** [comment]
Ugh. Trailing whitespace on a line.

@bors-servo try

**Comment by notriddle** [comment]
It got stuck at the top, again.

**Comment by frivoal** [comment]
Hello. I'm the (co)editor of the [CSS UI](https://drafts.csswg.org/css-ui-3/) spec where `text-overflow` is defined.

Are tests written for servo automatically submitted to either [Web Platform Tests](https://github.com/w3c/web-platform-tests/) or [the CSSWG's test repo](https://github.com/w3c/csswg-test/)? If yes, do you have a link to documentation about how that happens? If not, I would really appreciate if you could submit the test you wrote for this fix to [the CSSWG's test repo](https://github.com/w3c/csswg-test/). I'll merge the pull request as soon as you make it.

For the record (and so that we don't have to do another round of review once the pull request is made), I've reviewed the test, and the functional part of it is good. On the meta-data / formatting side, there are a couple of tweaks that should be made to match that repo's coding style guidelines, but I am happy to make them myself when you submit if you don't want to be bothered with that:

- Spec link: `<link rel="help" href="https://www.w3.org/TR/css-ui-3/#text-overflow">` should be added
- Naming: the file should end with `-001.html` (and `-001-ref.html`)
- Flags: `<meta name="flags" content="">` should be added
- Description: What you have as a title and as a description meta is what we'd expect in the `<meta name="assert" content="...">` meta. The title itself could be something like "CSS-UI test: text-overflow and reflow", and "CSS UI test reference" for the ref file.

**Comment by emilio** [comment]
Hi @frivoal,

I think they're not automatically synced, though that's something we probably want to do eventually. In any case, @Ms2ger, @jdm or @jgraham know more about the status of that. I'm aware that quite a lot of written tests in the `tests/wpt/mozilla` directory should be tidied up and upstreamed, but nobody has gotten to do it though.

Regarding that specific test, do you have cycles to upstream it @notriddle? Otherwise I can probably tidy it up myself.

On a kind of unrelated note, we should probably pass the automatic lints that wpt and csswg have on new tests, I'll fill an issue to do that, so tests can be more easily upstreamed.

Thanks again for writing.

**Comment by emilio** [comment]
I've filled #15058, and #15057 for the upstreaming and linting, respectively.

**Comment by notriddle** [comment]
I think I could upstream it, yeah. I wasn't sure if it would be wanted, since it ends in an identical DOM; it's literally checking the correctness of the incremental layout system.

**Comment by frivoal** [comment]
If Servo could get it wrong, so could others. I am not sure about this particular case, but I have seen incremental reflow bugs in multiple engines, so better be safe than sorry. More tests are always better :)

**Comment by notriddle** [comment]
Okay.

**Pull request merged**